### PR TITLE
Initial implementation of JWT:

### PR DESCRIPTION
**Pull request by jdrossl**
- Replace all authentication related code with Spring Security
- Implement HTTP Basic & JWT auth for the API
- Add new API for token management

https://github.com/craftercms/craftercms/issues/4429

Depends on https://github.com/craftercms/commons/pull/255

### CODE REVIEW

**Review by avasquez614 — APPROVED** [review]

File: src/main/java/org/craftercms/studio/impl/v2/security/listener/AuditLoginListener.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    public AuditLoginListener(StudioConfiguration studioConfiguration, SiteService siteService,
+                              AuditServiceInternal auditServiceInternal) {
+        super(studioConfiguration, siteService, auditServiceInternal);
+    }
+
+    @EventListener
+    public void recordAuthenticationSuccess(AuthenticationSuccessEvent event) {
+        if (event.getAuthentication() instanceof PreAuthenticatedAuthenticationToken) {
+            //TODO: Should we also record this events?
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/security/listener/AuditLoginListener.java]: Please log.

File: src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+
+    @Override
+    public String getUsername(String token) {
+        try {
+            var jwtConsumer = new JwtConsumerBuilder()
+                    .setEnableRequireEncryption()
+                    .setRequireSubject()
+                    .setExpectedIssuers(true, validIssuers)
+                    .setExpectedAudience(getActualAudience())
+                    .setRelaxVerificationKeyValidation() // TODO: Check if this is ok
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java]: Please remove per our discussion.

File: src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        if (id != null) {
+            claims.setJwtId(id.toString());
+        }
+
+        try {
+            // Sign the JWT
+            var jws = new JsonWebSignature();
+            jws.setPayload(claims.toJson());
+            jws.setKey(jwtSignKey);
+            jws.setAlgorithmHeaderValue(AlgorithmIdentifiers.HMAC_SHA512); // TODO: Choose an algorithm
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java]: Remove TODO

File: src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+            jws.setDoKeyValidation(false); // TODO: Check if this is ok
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java]: Remove.

File: src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+            // Encrypt the JWS
+            var jwe = new JsonWebEncryption();
+            jwe.setPayload(jws.getCompactSerialization());
+            // TODO: Choose algorithms
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v2/service/security/internal/AccessTokenServiceInternalImpl.java]: Remove todo

**Review by sumerjabri — CHANGES_REQUESTED** [review]
Please address the notes.

File: src/main/webapp/default-site/scripts/rest/api/1/security/login.post.groovy [hunk 6]
```diff
@@ -14,45 +14,8 @@
...
-        result.message = "Unauthorized"
-    } catch (Exception e) {
-        response.setStatus(500)
-        result.message = "Internal server error: \n" + e
-    }
-} catch (JsonException e) {
-    response.setStatus(400)
-    result.message = "Bad Request"
-}
+result.message = "This API has been deprecated"
```
Comment (sumerjabri) [inline src/main/webapp/default-site/scripts/rest/api/1/security/login.post.groovy]: This API has been replaced by a new login mechanism via JWT, please review the documentation.

**Review by sumerjabri — APPROVED** [review]

**Pull request merged**